### PR TITLE
CORTX-23443: adding retry decorator in consul confstore

### PR DESCRIPTION
**Pull request by rahul-tripathi-git**
Signed-off-by: Rahul Tripathi <rahul.tripathi@seagate.com>

# Problem Statement
- Consul based confstore was getting failed incase of some network issues, with HTTPException or ConsulException etc. Hence a retry logic is added with default 5 retries every time it fails.

# Design
-  Added a common decorator 'retry' in common.common module, It can be used wherever retry decorator is needed.

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing 
- [ ] Confirm that Test Cases are added (for both the cases, fix and feature) [Y/N]: 
- [ ] Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: 
- [x] Confirm Testing was performed with installed RPM [Y/N]:  

# Review Checklist 
  Before posting the PR please ensure
- [x] PR is self reviewed
- [ ] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [ ] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

File: py-utils/src/utils/kv_store/kv_store_collection.py [hunk 1]
```diff
@@ -436,6 +437,7 @@ def __init__(self, consul: Consul, store_path: str = '', delim: str = '>'):
             if not self._store_path.endswith('/'): self._store_path = self._store_path + '/'
         self._keys = self.get_keys()
 
+    @retry((ConsulException, HttpError, RequestException))
```
Comment (ujjwalpl) [inline py-utils/src/utils/kv_store/kv_store_collection.py:440]: JSon Implementation should not refer to Consul at all. This is clear violation of the design principle. .

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/kv_store/kv_store_collection.py (hunk 1 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/kv_store/kv_store_collection.py]: Hi @ujjwalpl 
This change is just for ConsulKV payload and not JSON implementation.

File: py-utils/src/utils/common/common.py [hunk 2]
```diff
@@ -4,6 +4,37 @@
 from cortx.utils.conf_store import Conf
 from cortx.utils.conf_store.error import ConfError
 
+def retry(ExceptionToCheck, tries=4, delay=3, backoff=2):
```
Comment (ujjwalpl) [inline py-utils/src/utils/common/common.py:7]: 1. Lets avoid global functions. 
2. Also, let caller decide the retry values. 
3. Current delay pattern would be 3s, 6s, 12s, 24s. If system is ready in say 10s, then you would wait for 21s.
    Normally delay needs to be smaller and retry count needs to be bigger, e.g. retries=6, delay=1, backoff=2. 
    Delay pattern would be 1, 2, 4, 8, 16, 32s. If system is ready 10s, then we would wait for 15s.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/common/common.py [hunk 3]
```diff
@@ -1,9 +1,41 @@
...
+
+    def __call__(self, func):
+        if len(func) == 1:
+            print(self._tries, self._exception)
+            print (func)
+            func = func[0]
+            def deco_retry(func):
+
+                @wraps(func)
+                def wrap(*x,**y):
```
Comment (selvakumaar5496) [inline py-utils/src/utils/common/common.py:25]: space after comma

File: py-utils/src/utils/common/common.py [hunk 4]
```diff
@@ -1,9 +1,41 @@
...
+                    mtries, mdelay = self._tries, self._delay
+                    while mtries > 1:
+                        try:
+                            return func(*x,**y)
```
Comment (selvakumaar5496) [inline py-utils/src/utils/common/common.py:29]: space after comma

File: py-utils/src/utils/common/common.py [hunk 5]
```diff
@@ -1,9 +1,41 @@
...
+                        except self._exception:
+                            time.sleep(mdelay)
+                            mtries -= 1
+                            mdelay *= self._backoff
+                    return func(*x,**y)
```
Comment (selvakumaar5496) [inline py-utils/src/utils/common/common.py:34]: space after comma

File: py-utils/src/utils/common/common.py [hunk 6]
```diff
@@ -1,9 +1,41 @@
...
+                return wrap
+            return deco_retry
+        else:
+            raise UtilsError(errno.EINVAL, "Invalid arguments passed to Retry decorator", '')
```
Comment (selvakumaar5496) [inline py-utils/src/utils/common/common.py:38]: why do we have hardcoded kwargs?

File: py-utils/src/utils/common/common.py [hunk 7]
```diff
@@ -1,9 +1,41 @@
...
+
+                @wraps(func)
+                def wrap(*x,**y):
+                    mtries, mdelay = self._tries, self._delay
+                    while mtries > 1:
+                        try:
+                            return func(*x,**y)
+                        except self._exception:
+                            time.sleep(mdelay)
+                            mtries -= 1
```
Comment (rohit-k-dwivedi) [inline py-utils/src/utils/common/common.py:32]: mtries and mdelay can be renames to better pronounceable variables, I guess m here stands for max?
max_tries and max_delay

Comment (suryakumar1024) [reply py-utils/src/utils/common/common.py:32]: @rohit-k-dwivedi good suggestion in that case

File: py-utils/src/utils/common/common.py (hunk 3 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py]: Done

File: py-utils/src/utils/common/common.py (hunk 4 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py]: Done

File: py-utils/src/utils/common/common.py (hunk 5 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py]: Done

File: py-utils/src/utils/common/common.py (hunk 6 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py:36]: We have hardcoded kwargs to set default values,

File: py-utils/src/utils/common/common.py (hunk 7 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py]: Done

Comment (rohit-k-dwivedi) [reply py-utils/src/utils/common/common.py]: please take care of *x and **y too.

Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py]: replaced it with *args and **kwargs

**Comment by rahul-tripathi-git** [comment]
Codacy issue mentioned is wrong, ConsulException is imported in line 21 of py-utils/src/utils/kv_store/kv_store_collection.py.

File: py-utils/src/utils/common/common.py [hunk 8]
```diff
@@ -1,9 +1,52 @@
...
+        self._delay = delay
+        self._backoff = backoff
+
+    def __call__(self, func):
+        @wraps(func)
+        def wrap(*args,**kwargs):
+            max_tries, max_delay = self._tries, self._delay
+            while max_tries > 1:
+                try:
+                    return func(*args,**kwargs)
```
Comment (ujjwalpl) [inline py-utils/src/utils/common/common.py:43]: `return func(*args, **kwargs)`

File: py-utils/src/utils/common/common.py [hunk 9]
```diff
@@ -1,9 +1,52 @@
 import os
 import errno
+import time
+from functools import wraps
 
 from cortx.utils.conf_store import Conf
 from cortx.utils.conf_store.error import ConfError
 
+class Retries:
```
Comment (ujjwalpl) [inline py-utils/src/utils/common/common.py:9]: Few suggestions on the names...
1. ExponentialBackoff 
2. RetryBackoff
3. RetryCount

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by selvakumaar5496 — APPROVED** [review]

File: py-utils/src/utils/common/common.py [hunk 10]
```diff
@@ -1,9 +1,52 @@
...
+
+    def __call__(self, func):
+        @wraps(func)
+        def wrap(*args,**kwargs):
+            max_tries, max_delay = self._tries, self._delay
+            while max_tries > 1:
+                try:
+                    return func(*args,**kwargs)
+                except self._exception:
+                    time.sleep(max_delay)
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py]: Done.

File: py-utils/src/utils/kv_store/kv_store_collection.py [hunk 11]
```diff
@@ -451,14 +453,18 @@ def get(self, key: str, *args, **kwargs) -> str:
             raise KvError(errno.EINVAL, \
                 "Invalid response from consul: %d:%s", index, data)
 
+    @Retries(exception=(ConsulException, HttpError, RequestException), tries=4, delay=1, backoff=2)
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply py-utils/src/utils/kv_store/kv_store_collection.py:456]: Done

File: py-utils/src/utils/common/common.py (hunk 8 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py:45]: Done

File: py-utils/src/utils/common/common.py (hunk 9 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py]: Done. Renamed to ExponentialBackoff.

**Review by ujjwalpl — APPROVED** [review]

File: py-utils/src/utils/common/common.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/common.py]: Added a cap argument.
modified the sleep logis as -> sleep = min( cap, delay*backoff)

File: py-utils/src/utils/common/common.py [hunk 12]
```diff
@@ -1,9 +1,55 @@
...
+    or
+    @ExponentialBackoff(Exception, tries=2, delay=10, backoff=2, cap=2)
+    def bar():
+        ...
+
+    Args:
+        exception (Exception or tuple(Exception)):  it can be a single exception
+                                                    or a tuple of exceptions.
+        tries (int): 10     number of times to try before failing.
+        delay (int): 1      initiall delay (in seconds) between retries.
```
Comment (cdeshmukh) [inline py-utils/src/utils/common/common.py:28]: typo initiall -> initial

File: py-utils/src/utils/common/common.py [hunk 13]
```diff
@@ -1,9 +1,55 @@
 import os
 import errno
+import time
+from functools import wraps
 
 from cortx.utils.conf_store import Conf
 from cortx.utils.conf_store.error import ConfError
 
+class ExponentialBackoff:
```
Comment (cdeshmukh) [inline py-utils/src/utils/common/common.py:9]: This is interesting and unique. 
However latency for all these calls would change, these should be communicated to teams now

**Pull request merged**